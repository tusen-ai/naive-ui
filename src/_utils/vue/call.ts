type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
