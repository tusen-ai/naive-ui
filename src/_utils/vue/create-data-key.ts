export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
