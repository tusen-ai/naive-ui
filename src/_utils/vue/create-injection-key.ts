import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
