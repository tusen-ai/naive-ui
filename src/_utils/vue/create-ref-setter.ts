import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
