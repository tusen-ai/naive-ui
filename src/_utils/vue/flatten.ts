import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
