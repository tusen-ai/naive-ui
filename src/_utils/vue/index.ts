export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
