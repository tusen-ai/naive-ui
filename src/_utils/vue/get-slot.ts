import { type ComponentPublicInstance, type VNodeChild } from 'vue'

export function getSlot (
  instance: ComponentPublicInstance,
  slotName = 'default',
  fallback: VNodeChild[] = []
): VNodeChild[] {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
