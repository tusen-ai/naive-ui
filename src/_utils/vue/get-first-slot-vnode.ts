import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
