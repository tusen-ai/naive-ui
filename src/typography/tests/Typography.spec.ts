import { h } from 'vue'
import { mount } from '@vue/test-utils'
import {
  NA,
  NBlockquote,
  NH1,
  NH2,
  NH3,
  NH4,
  NH5,
  NH6,
  NLi,
  NOl,
  NP,
  NText,
  NUl
} from '../index'

describe('n-a', () => {
  it('should work with import on demand', () => {
    mount(NA)
  })

  it('should work with normal', () => {
    const wrapper = mount(NA, { props: { href: '/test' } })

    expect(wrapper.find('a').classes()).toContain('n-a')
    expect(wrapper.find('a').attributes('href')).toBe('/test')
  })
})

describe('n-text', () => {
  it('should work with import on demand', () => {
    mount(NText)
  })

  it('should work with `type` prop', async () => {
    const wrapper = mount(NText, { slots: { default: () => 'test' } })

    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ type: 'info' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ type: 'success' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ type: 'warning' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ type: 'error' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()
  })

  it('should work with Font style', () => {
    let wrapper

    wrapper = mount(NText, {
      props: { strong: true },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').classes()).toContain('n-text--strong')

    wrapper = mount(NText, {
      props: { italic: true },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').classes()).toContain('n-text--italic')

    wrapper = mount(NText, {
      props: { underline: true },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').classes()).toContain('n-text--underline')

    wrapper = mount(NText, {
      props: { delete: true },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').classes()).toContain('n-text--delete')

    wrapper = mount(NText, {
      props: { code: true },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').classes()).toContain('n-text--code')
  })

  it('should work with `depth` prop', async () => {
    const wrapper = mount(NText, { slots: { default: () => 'test' } })

    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ depth: '1' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ depth: '2' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ depth: '3' })
    expect(wrapper.find('.n-text').attributes('style')).toMatchSnapshot()
  })

  it('should work with `tag` prop', async () => {
    const wrapper = mount(NText, {
      props: { tag: 'div' },
      slots: { default: () => 'test' }
    })
    expect(wrapper.find('.n-text').html()).toContain('test</div>')
  })
})

describe('n-p', () => {
  it('should work with import on demand', () => {
    mount(NP)
  })

  it('should work with `depth` prop', async () => {
    const wrapper = mount(NP, { slots: { default: () => 'test' } })

    expect(wrapper.find('p').classes()).toContain('n-p')

    await wrapper.setProps({ depth: '1' })
    expect(wrapper.find('.n-p').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ depth: '2' })
    expect(wrapper.find('.n-p').attributes('style')).toMatchSnapshot()

    await wrapper.setProps({ depth: '3' })
    expect(wrapper.find('.n-p').attributes('style')).toMatchSnapshot()
  })
})

describe('n-h1 n-h2 n-h3 n-h4 n-h5 n-h6', () => {
  it('should work with import on demand', () => {
    mount(NH1)
    mount(NH2)
    mount(NH3)
    mount(NH4)
    mount(NH5)
    mount(NH6)
  })

  it('should work with normal', async () => {
    let wrapper: any

    wrapper = mount(NH1, {
      props: { prefix: 'bar', alignText: true },
      slots: { default: () => 'test' }
    })
    ;(['n-h1', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()

    wrapper = mount(NH2, {
      props: { prefix: 'bar', alignText: true, type: 'default' },
      slots: { default: () => 'test' }
    })
    ;(['n-h2', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()

    wrapper = mount(NH3, {
      props: { prefix: 'bar', alignText: true, type: 'success' },
      slots: { default: () => 'test' }
    })
    ;(['n-h3', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()

    wrapper = mount(NH4, {
      props: { prefix: 'bar', alignText: true, type: 'info' },
      slots: { default: () => 'test' }
    })
    ;(['n-h4', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()

    wrapper = mount(NH5, {
      props: { prefix: 'bar', alignText: true, type: 'warning' },
      slots: { default: () => 'test' }
    })
    ;(['n-h5', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()

    wrapper = mount(NH6, {
      props: { prefix: 'bar', alignText: true, type: 'error' },
      slots: { default: () => 'test' }
    })
    ;(['n-h6', 'n-h--prefix-bar', 'n-h--align-text'] as const).forEach(
      (item) => {
        expect(wrapper.find('.n-h').classes()).toContain(item)
      }
    )
    expect(wrapper.find('.n-h').attributes('style')).toMatchSnapshot()
  })
})

describe('n-ul n-ol n-li', () => {
  it('should work with import on demand', () => {
    mount(NUl)
    mount(NOl)
    mount(NLi)
  })

  it('should work with normal', () => {
    let wrapper = mount(NUl, {
      props: { alignText: true },
      slots: {
        default: () =>
          h(NLi, null, {
            default: () => 'test'
          })
      }
    })
    ;(['n-ul', 'n-ul--align-text'] as const).forEach((item) => {
      expect(wrapper.find('ul').classes()).toContain(item)
    })
    expect(wrapper.find('li').text()).toBe('test')
    expect(wrapper.find('ul').attributes('style')).toMatchSnapshot()

    wrapper = mount(NOl, {
      props: { alignText: true },
      slots: {
        default: () =>
          h(NLi, null, {
            default: () => 'test'
          })
      }
    })
    ;(['n-ol', 'n-ol--align-text'] as const).forEach((item) => {
      expect(wrapper.find('ol').classes()).toContain(item)
    })
    expect(wrapper.find('li').text()).toBe('test')
    expect(wrapper.find('ol').attributes('style')).toMatchSnapshot()
  })
})

describe('n-blockquote', () => {
  it('should work with import on demand', () => {
    mount(NBlockquote)
  })

  it('should work with normal', () => {
    const wrapper = mount(NBlockquote, {
      props: { alignText: true },
      slots: { default: () => 'test' }
    })

    ;(['n-blockquote', 'n-blockquote--align-text'] as const).forEach((item) => {
      expect(wrapper.find('blockquote').classes()).toContain(item)
    })
    expect(wrapper.find('blockquote').attributes('style')).toMatchSnapshot()
  })
})
