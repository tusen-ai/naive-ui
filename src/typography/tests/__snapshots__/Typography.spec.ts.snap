// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`n-blockquote > should work with normal 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-prefix-color: rgb(224, 224, 230); --n-text-color: rgb(51, 54, 57);"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 30px; --n-margin: 28px 0 20px 0; --n-bar-color: #18a058; --n-bar-width: 4px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 16px;"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 22px; --n-margin: 28px 0 20px 0; --n-bar-color: #18a058; --n-bar-width: 4px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 16px;"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 18px; --n-margin: 28px 0 20px 0; --n-bar-color: #18a058; --n-bar-width: 3px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 12px;"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 16px; --n-margin: 28px 0 18px 0; --n-bar-color: #2080f0; --n-bar-width: 3px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 12px;"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 5`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 16px; --n-margin: 28px 0 18px 0; --n-bar-color: #f0a020; --n-bar-width: 3px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 12px;"`;

exports[`n-h1 n-h2 n-h3 n-h4 n-h5 n-h6 > should work with normal 6`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 16px; --n-margin: 28px 0 18px 0; --n-bar-color: #d03050; --n-bar-width: 3px; --n-font-weight: 500; --n-text-color: rgb(31, 34, 37); --n-prefix-width: 12px;"`;

exports[`n-p > should work with \`depth\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-margin: 16px 0 16px 0; --n-text-color: rgb(31, 34, 37);"`;

exports[`n-p > should work with \`depth\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-margin: 16px 0 16px 0; --n-text-color: rgb(51, 54, 57);"`;

exports[`n-p > should work with \`depth\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-margin: 16px 0 16px 0; --n-text-color: rgb(118, 124, 130);"`;

exports[`n-text > should work with \`depth\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(51, 54, 57); --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`depth\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(31, 34, 37); --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`depth\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(51, 54, 57); --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`depth\` prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(118, 124, 130); --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`type\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(51, 54, 57); --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`type\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: #2080f0; --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`type\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: #18a058; --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`type\` prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: #f0a020; --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-text > should work with \`type\` prop 5`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: #d03050; --n-font-weight-strong: 500; --n-font-famliy-mono: v-mono, SFMono-Regular, Menlo, Consolas, Courier, monospace; --n-code-border-radius: 2px; --n-code-text-color: rgb(51, 54, 57); --n-code-color: rgb(244, 244, 248); --n-code-border: 1px solid #0000;"`;

exports[`n-ul n-ol n-li > should work with normal 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-text-color: rgb(51, 54, 57); --n-li-margin: .25em 0 0 0; --n-ol-padding: 0 0 0 2em; --n-ul-padding: 0 0 0 2em;"`;

exports[`n-ul n-ol n-li > should work with normal 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-font-size: 14px; --n-line-height: 1.6; --n-text-color: rgb(51, 54, 57); --n-li-margin: .25em 0 0 0; --n-ol-padding: 0 0 0 2em; --n-ul-padding: 0 0 0 2em;"`;
